### PR TITLE
RavenDB-19100 - Have an EP to modify the database topology

### PR DESCRIPTION
**Pull request by shaharhikri**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19100/Have-an-EP-to-modify-the-database-topology

### Additional description

Have an EP to modify the database topology

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Client/ServerWide/DatabaseTopology.cs [hunk 1]
```diff
@@ -212,6 +212,57 @@ public bool RelevantFor(string nodeTag)
                    Promotables.Contains(nodeTag);
         }
 
+        internal DatabaseTopology CopyTo(DatabaseTopology databaseTopology)
```
Comment (karmeli87) [inline src/Raven.Client/ServerWide/DatabaseTopology.cs]: why do we need that?

File: src/Raven.Server/Web/System/DatabasesHandler.cs [hunk 2]
```diff
@@ -83,6 +86,83 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
...
 
+        [RavenAction("/topology/modify", "POST", AuthorizationStatus.ValidUser, EndpointType.Read)]
+        public async Task ModifyTopology()
+        {
+            var dbName = GetQueryStringValueAndAssertIfSingleAndNotEmpty("name").Trim();
+            var raftRequestId = GetRaftRequestIdFromQuery();
+
+            await ServerStore.EnsureNotPassiveAsync();
+            using (ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            using (context.OpenReadTransaction())
```
Comment (karmeli87) [inline src/Raven.Server/Web/System/DatabasesHandler.cs]: let's avoid keep a read tx here.

File: src/Raven.Server/Web/System/DatabasesHandler.cs [hunk 3]
```diff
@@ -83,6 +86,83 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
             }
         }
 
+        [RavenAction("/topology/modify", "POST", AuthorizationStatus.ValidUser, EndpointType.Read)]
```
Comment (karmeli87) [inline src/Raven.Server/Web/System/DatabasesHandler.cs]: This should `Operator` clearance.

Comment (karmeli87) [reply src/Raven.Server/Web/System/DatabasesHandler.cs]: and remove `EndpointType`

File: src/Raven.Server/Web/System/DatabasesHandler.cs [hunk 4]
```diff
@@ -83,6 +86,83 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
...
+
+                        var url = clusterTopology.GetUrlFromTag(node);
+                        if (url == null)
+                            throw new InvalidOperationException($"Can't modify database {dbName} topology, because node {node} (which is in the new topology) is not part of the cluster");
+
+                        if (databaseRecord.Encrypted && url.StartsWith("https:", StringComparison.OrdinalIgnoreCase) == false)
+                            throw new InvalidOperationException($"Can't modify database {dbName} topology, because database {dbName} is encrypted but node {node} (which is in the new topology) doesn't have an SSL certificate.");
+                    }
+                }
+                databaseTopology.ReplicationFactor = Math.Min(databaseTopology.Count, clusterTopology.AllNodes.Count);
```
Comment (karmeli87) [inline src/Raven.Server/Web/System/DatabasesHandler.cs]: This looks like a code duplication from `AddDatabaseNode`, is that possible to reuse?

File: test/SlowTests/Issues/RavenDB-19100.cs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        {
+            var (nodes, leader) = await CreateRaftCluster(3);
+            var storeOptions = new Options { Server = leader, ReplicationFactor = nodes.Count };
+
+            using (var store = GetDocumentStore(storeOptions))
+            {
+                var databaseName = store.Database;
+                var record = await store.Maintenance.Server.SendAsync(new GetDatabaseRecordOperation(databaseName));
+                var databaseTopology = new DatabaseTopology();
+                record.Topology.CopyTo(databaseTopology);
```
Comment (karmeli87) [inline test/SlowTests/Issues/RavenDB-19100.cs]: I think you can modify the topology here directly

**Comment by karmeli87** [comment]
Note, for shrading we probably need to have the ability to modify specific shard and modify the orchestrator topology

File: src/Raven.Server/Web/System/DatabasesHandler.cs (hunk 4 above)
Comment (shaharhikri) [reply src/Raven.Server/Web/System/DatabasesHandler.cs]: Only 3 validation checks are the same.
Also, the exception messages are different.

File: src/Raven.Server/Web/System/DatabasesHandler.cs (hunk 2 above)
Comment (shaharhikri) [reply src/Raven.Server/Web/System/DatabasesHandler.cs]: Cant do that because of this line:
https://github.com/ravendb/ravendb/blob/b13fe40dc9e53ae42121aa52333f8d2255bc4fcd/src/Raven.Server/Web/System/DatabasesHandler.cs#L111

File: test/SlowTests/Issues/RavenDB-19100.cs (hunk 5 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-19100.cs]: Done

File: src/Raven.Server/Web/System/DatabasesHandler.cs (hunk 3 above)
Comment (shaharhikri) [reply src/Raven.Server/Web/System/DatabasesHandler.cs]: Done

File: src/Raven.Client/ServerWide/DatabaseTopology.cs (hunk 1 above)
Comment (shaharhikri) [reply src/Raven.Client/ServerWide/DatabaseTopology.cs]: Done

File: src/Raven.Server/Web/System/DatabasesHandler.cs (hunk 4 above)
Comment (shaharhikri) [reply src/Raven.Server/Web/System/DatabasesHandler.cs]: Done

File: src/Raven.Server/Web/System/DatabasesHandler.cs (hunk 2 above)
Comment (shaharhikri) [reply src/Raven.Server/Web/System/DatabasesHandler.cs]: Done

**Review by karmeli87 — APPROVED** [review]

**Comment by karmeli87** [comment]
@shaharhikri please check failing tests

File: src/Raven.Server/Web/System/DatabasesHandler.cs [hunk 6]
```diff
@@ -83,6 +87,74 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
...
+                await ServerStore.Cluster.WaitForIndexNotification(newIndex);
+
+                // Return Raft Index
+                HttpContext.Response.StatusCode = (int)HttpStatusCode.OK;
+
+                await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBodyStream()))
+                {
+                    context.Write(writer, new DynamicJsonValue
+                    {
+                        [nameof(DatabasePutResult.RaftCommandIndex)] = newIndex
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/DatabasesHandler.cs]: wrong nameof, should be ModifyDatabaseTopologyResult

File: src/Raven.Server/Web/System/DatabasesHandler.cs [hunk 7]
```diff
@@ -83,6 +87,74 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
...
+                {
+                    Topology = databaseTopology
+                };
+
+                var (newIndex, _) = await ServerStore.SendToLeaderAsync(update);
+
+                await ServerStore.Cluster.WaitForIndexNotification(newIndex);
+
+                // Return Raft Index
+                HttpContext.Response.StatusCode = (int)HttpStatusCode.OK;
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/DatabasesHandler.cs]: not needed, this is a default one

File: src/Raven.Server/Web/System/DatabasesHandler.cs [hunk 8]
```diff
@@ -83,6 +87,74 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
...
+
+                // Update Topology
+                var update = new UpdateTopologyCommand(dbName, SystemTime.UtcNow, raftRequestId)
+                {
+                    Topology = databaseTopology
+                };
+
+                var (newIndex, _) = await ServerStore.SendToLeaderAsync(update);
+
+                await ServerStore.Cluster.WaitForIndexNotification(newIndex);
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/DatabasesHandler.cs]: we are waiting here for cluster, if database is loaded, should we wait for it as well? cc @karmeli87

File: src/Raven.Server/Web/System/DatabasesHandler.cs [hunk 9]
```diff
@@ -83,6 +87,74 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
...
+                }
+
+                // Validate Database Name
+                DatabaseRecord databaseRecord;
+                using (context.OpenReadTransaction())
+                    databaseRecord = ServerStore.Cluster.ReadDatabase(context, dbName, out var index);
+
+                if (databaseRecord == null)
+                {
+                    throw new DatabaseDoesNotExistException($"Database {dbName} Record not found when trying to modify database topology");
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/DatabasesHandler.cs]: there is a static method called `Throw` or 'ThrowWithMessage' in that exception, please use it.

File: src/Raven.Server/Web/System/DatabasesHandler.cs [hunk 10]
```diff
@@ -83,6 +87,74 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
...
+            {
+                var json = await context.ReadForDiskAsync(RequestBodyStream(), "Database Topology");
+                var databaseTopology = JsonDeserializationCluster.DatabaseTopology(json);
+
+                if (LoggingSource.AuditLog.IsInfoEnabled)
+                {
+                    var clientCert = GetCurrentCertificate();
+
+                    var auditLog = LoggingSource.AuditLog.GetLogger("DbMgmt", "Audit");
+                    auditLog.Info($"Database \'{dbName}\' record has been modified by {clientCert?.Subject} ({clientCert?.Thumbprint})");
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/DatabasesHandler.cs]: this is very inaccurate
1. It is in a past tense
2. there is no info here regarding the modify topology, please match message pattern in other places that use this

File: src/Raven.Server/Web/RequestHandler.cs [hunk 11]
```diff
@@ -159,6 +159,24 @@ private static bool HeadersAllowGzip(IDictionary<string, Microsoft.Extensions.Pr
             return false;
         }
 
+        protected void ValidateNodeForAddingToDb(string databaseName, string node, DatabaseRecord databaseRecord, ClusterTopology clusterTopology, string baseMessage = null)
```
Comment (ppekrol) [inline src/Raven.Server/Web/RequestHandler.cs]: I would prefer this to be a static method somewhere. DatabaseHelper?

File: src/Raven.Server/Web/System/DatabasesHandler.cs (hunk 8 above)
Comment (shaharhikri) [reply src/Raven.Server/Web/System/DatabasesHandler.cs]: remove ```await ServerStore.Cluster.WaitForIndexNotification(newIndex);``` and wait for the raft index in the tests instead?

File: src/Raven.Server/Web/RequestHandler.cs (hunk 11 above)
Comment (shaharhikri) [reply src/Raven.Server/Web/RequestHandler.cs]: Done

File: src/Raven.Server/Web/System/DatabasesHandler.cs (hunk 10 above)
Comment (shaharhikri) [reply src/Raven.Server/Web/System/DatabasesHandler.cs]: Done

File: src/Raven.Server/Web/System/DatabasesHandler.cs (hunk 9 above)
Comment (shaharhikri) [reply src/Raven.Server/Web/System/DatabasesHandler.cs]: Done

File: src/Raven.Server/Web/System/DatabasesHandler.cs (hunk 7 above)
Comment (shaharhikri) [reply src/Raven.Server/Web/System/DatabasesHandler.cs]: Done

File: src/Raven.Server/Web/System/DatabasesHandler.cs (hunk 6 above)
Comment (shaharhikri) [reply src/Raven.Server/Web/System/DatabasesHandler.cs]: Done

File: src/Raven.Server/Web/System/DatabasesHandler.cs (hunk 8 above)
Comment (ppekrol) [reply src/Raven.Server/Web/System/DatabasesHandler.cs]: no, using databasesLandlord you can checkk if database is loaded and then get it from there, afterwards you can wait for the command to be applied

File: src/Raven.Server/Web/System/DatabasesHandler.cs [hunk 12]
```diff
@@ -83,6 +87,77 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
             }
         }
 
+        [RavenAction("/topology/modify", "POST", AuthorizationStatus.Operator)]
```
Comment (karmeli87) [inline src/Raven.Server/Web/System/DatabasesHandler.cs]: @ppekrol I think we need to change the URL here... 
What do you think about `/admin/databases/topology/modify`?

File: src/Raven.Server/Web/System/DatabasesHandler.cs (hunk 8 above)
Comment (karmeli87) [reply src/Raven.Server/Web/System/DatabasesHandler.cs]: We don't wait for it when we adding a new node, so I would argue that we don't need to wait for it here as well

File: test/SlowTests/Authentication/AuthenticationBasicTests.cs [hunk 13]
```diff
@@ -401,7 +401,8 @@ public void AllAdminAuthorizationStatusHaveCorrectRoutes()
             {
                 "/monitoring/snmp/oids",
                 "/monitoring/snmp",
-                "/monitoring/snmp/bulk"
+                "/monitoring/snmp/bulk",
+                "/topology/modify"
```
Comment (karmeli87) [inline test/SlowTests/Authentication/AuthenticationBasicTests.cs]: why is that needed here? looks wired to me it is the only EP of that type here..

File: test/SlowTests/Issues/RavenDB-19100.cs [hunk 14]
```diff
@@ -0,0 +1,118 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using FastTests;
+using FastTests.Client;
+using NuGet.ContentModel;
```
Comment (karmeli87) [inline test/SlowTests/Issues/RavenDB-19100.cs]: ?

File: test/SlowTests/Authentication/AuthenticationBasicTests.cs (hunk 13 above)
Comment (shaharhikri) [reply test/SlowTests/Authentication/AuthenticationBasicTests.cs]: Removed

File: test/SlowTests/Issues/RavenDB-19100.cs (hunk 14 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-19100.cs]: Removed

File: src/Raven.Server/Web/System/DatabasesHandler.cs (hunk 12 above)
Comment (shaharhikri) [reply src/Raven.Server/Web/System/DatabasesHandler.cs]: Done

**Comment by karmeli87** [comment]
test this please
